test/datahike/test/validation_test.cljc

```diff
@@ -30,7 +30,7 @@
       [{:db/id -1 :profile #"regexp"}])
 
     (is (thrown-with-msg? Throwable #"Unknown operation" (d/db-with db [["aaa" :name "Ivan"]])))
-    (is (thrown-with-msg? Throwable #"Bad entity type at" (d/db-with db [:db/add "aaa" :name "Ivan"])))
+    #_(is (thrown-with-msg? Throwable #"Bad entity type at" (d/db-with db [:db/add "aaa" :name "Ivan"]))) ;; todo: how to test incorrect spec now?
```

Comment:
At the moment the specs are being checked when the tests are run, so `db-with` already fails on the specs and it doesn't even get to the point where it would throw "Bad entity type". 

The question is now if we want to to keep on checking the specs for testing or we should  be able to have tests like this one here, which is good to have for documentation purposes in order to see what's happening.

Reply:
But if we were not using the specs in the tests the question would be where and how else we would use them. One option would be in development environments. Probably not production though as is would slow down the code, right?

Reply:
Yes, I think this is a common way to enforce contracts/spec like that, to activate them in development and debugging and turn them off in production. It really depends though also on the production environment, since they guard an interface to external access here. The main direct benefit of having specs at all in my opinion is to have automatically checked documentation for what the inputs should be.

Reply:
I think this here is an edge case where we would have two guards. Ideally we could just tell that this spec check should be always on, not sure whether this is possible.

Reply:
Hm, I can't find any documentation about how it would be possible to automatically turn on instrumentation for development. It looks like it is always turned on with calling `s/instrument` in the repl. There are libraries that are doing it differently, but they might have less features. 

I would suggest removing the instrumentation for the tests for now and instead use the specs for generative testing in the future.

Reply:
A way could be to user `https://github.com/gfredericks/user.clj` together with the dev build and add `(t/instrument)` to the existing user file. I would anyway move to this file for experimental code from the old sandbox file.

Reply:
Did you look into [unstrument?](https://github.com/jeaye/orchestra/blob/master/src/clj/orchestra/spec/test.cljc#L51) Maybe we can stop the instrumentation for certain tests?

Reply:
```
(st/unstrument)
(is (thrown-with-msg? Throwable #"Bad entity type at" (d/db-with db [:db/add "aaa" :name "Ivan"])))
(st/instrument))
```
Seems to work. Don't know if this works well because of the concurrency of the tests, though.

Reply:
Hm, that's another option. I could live with that as well, but I think I would still personally prefer removing the instrumentation from the tests entirely.

Reply:
I am fine with not instrumenting the specs but then probably no one ever uses them again:) So maybe we should have some kind of interface-test? Just an idea...

Reply:
What do you mean by 'interface-test'? 

As mentioned above, I would like to use them for generative tests which would act like a check if the specs are correct as well as enable us to potentially discover bugs that we wouldn't have discovered otherwise. Is that what you mean as well?

Reply:
Yes, I'd say that's what I meant. But I would rephrase it a bit... the specs are basically defining our contract with the world and therefore we should keep some kind of test that this contract is not broken and works reliably.

Reply:
I have created issue #605 for this topic now. Feel free to change the wording or add stuff that you think is missing.